src/python/pants/testutil/rule_runner.py

```diff
@@ -656,89 +659,163 @@ class MockGet(Generic[_O]):
 
-    return (
-        f"HINT: missing : {[str(e) for e in missing]}\n"
-        f"HINT: additional : {[str(e) for e in additional]}\n"
-    )
+    if union_membership:
+        for mg in list(missing_gets):
+            for sg in list(actual_gets_as_comparable):
+                if all(
+                    (
+                        input_type in sg.input_types
+                        or (
+                            input_type in union_membership
+                            and any(
+                                union_membership.is_member(input_type, sgit)
+                                for sgit in sg.input_types
+                            )
+                        )
+                    )
+                    for input_type in mg.input_types
+                ):
+                    missing_gets.remove(mg)
+                    surplus_gets.discard(sg)
+                    break
+
+    errors = []
+    warnings = []
+
+    # See if missing calls can be satisfied by a MockGet.
+    for missing_call_name in list(missing_call_names):
+        missing_call = next(ec for ec in expected_calls if ec.rule_id == missing_call_name)
+        # NB: We don't have input types for by-name calls, so we can't use the "as_comparable"
+        #  to compare the missing call to actual gets. We heuristically compare just output types
+        #  instead. In some rare cases (where we have multiple mocked rules with the same
+        #  output type) this might warn when it should error, but it won't error when it shouldn't
+        #  (and of course the test will still fail as it should, we just won't have as obvious
+        #  an error message to show).
+        for ag in actual_gets_as_comparable:
+            if missing_call.output_type == ag.output_type:
+                warnings.append(
+                    f"A call to {missing_call} was provided by a mock_get and not by a "
+                    "mock_call. This works for now but will be deprecated soon."
+                )
+                missing_call_names.remove(missing_call_name)
+                surplus_gets.discard(ag)
+
+    if missing_call_names:
+        errors.append(f"Missing mocks for calls: {', '.join(missing_call_names)}")
+    if surplus_call_names:
+        errors.append(f"Surplus mocks for calls: {', '.join(surplus_call_names)}")
+    if missing_gets:
+        errors.append(f"Missing mocks for gets: {', '.join(str(e) for e in missing_gets)}")
+    if surplus_gets:
+        errors.append(f"Surplus mocks for gets: {', '.join(str(e) for e in surplus_gets)}")
+
+    return "\n".join(errors), "\n".join(warnings)
 
 
 def run_rule_with_mocks(
     rule: Callable[..., Coroutine[Any, Any, _O]],
     *,
     rule_args: Sequence[Any] = (),
     mock_gets: Sequence[MockGet | MockEffect] = (),
+    mock_calls: Mapping[str, Callable] | None = None,
     union_membership: UnionMembership | None = None,
+    warn_on_calls_satisfied_by_gets: bool = True,
```

Comment:
The tests of the `test` goal itself turn these warnings off so that they don't contaminate the output it checks for.